### ISSUE TITLE
<jsp-file>incorrectly handled when the jsp is at the top directory

### ISSUE THREAD

**Issue by ludoch** [eclipse/jetty.project#1268]
This does not seem to work:

 ```   
    <servlet>
        <servlet-name>ExportBeta</servlet-name>
        <jsp-file>/index.jsp</jsp-file>
    </servlet>
    <servlet-mapping>
        <servlet-name>ExportBeta</servlet-name>
        <url-pattern>/app/export</url-pattern>
    </servlet-mapping>
```
as the servlet class name used to map this jsp is 

org.apache.jsp..index_jsp

instead of

org.apache.jsp.index_jsp

in

return getJspPackagePrefix() + "." +getPackageOfJspClass(jsp) + "." + getNameOfJspClass(jsp);

getPackageOfJspClass can be empty

**Comment by joakime** [comment]
The upgrade to Jasper as part of issue #1205 might fix this too.

Need to dig into Jasper bug tracker.

**Comment by gregw** [comment]
I think we also need to make this more easily extensible

**Comment by janbartel** [comment]
Looks like a simple bug: we try to find out what the package of the class should be, we convert '/' in the jsp-file string to '.' and remove the name of the jsp from the end;  we don't handle the case where we're left with "" as the package name.

**Comment by janbartel** [comment]
@ludoch it is not essential to have the org.apache.jasper.compiler.JspUtil class: if it is present, we will use it to generate the proper package name. If it isn't there, we use the algorithm from my previous comment.

**Comment by ludoch** [comment]
@janbartel : yes, it seems simple to fix and thanks for the extra info...
@gregw: yes, it could be simpler to customize: we could use in the future a different package naming (in fact, we did for Jetty6 and for Jetty9 I kept the default, maybe because of this lack of easy customization..)

**Comment by janbartel** [comment]
Fix checked in for jetty-9.3 and above.

@ludoch  I fixed the problem with empty package name, and also fixed 2 other problems I discovered (that weren't affecting you), and I also changed the visibility of the getJspPackagePrefix(), getPackageOfJspClass(jsp), getNameOfJspClass(jsp) and getClassnameForJsp(jsp) methods to protected so you can override everything if you want.

**Comment by ludoch** [comment]
Curious about what would be the final error message for a customer? 
So far, it was an obscure 
javax.servlet.ServletContext log: unavailable
javax.servlet.UnavailableException: No class in holder

Could we have the class name somewhere so that we get a better debugging hint?
I did not see integration test to make sure we cover the test case of precompiled jsps..(maybe later?)

**Comment by janbartel** [comment]
The change I made means that if the jsp-file isn't convertible into a fully qualified classname, the ServletHolder held class remains as null. I've tweaked the messaging now. This is an example of the messaging to the user in this case:

LOG.WARN: Bad jsp-file /blah/ conversion to classname in holder foo
UnavailableException: No class in holder foo@18cc6==null,jsp=/blah/,order=-1,inst=false
  
The meaning of the various parts of the UnavailableException are:
 foo:   from servlet-name in web.xml (or annotation etc)
18cc6:  hashcode for ServletHolder
null:  the classname of the servlet
jsp=:  the jsp-file from web.xml
order=:  init-order from web.xml
inst=: only true if the ServletHolder is holding an instance of a servlet (always false in this case)

If the jsp-file _was_ able to be successfully converted into a fully qualified classname, we check for a precompiled jsp. If we can't find one, and there is a jsp servlet present, we set the classname to be the jsp servlet (eg org.eclipse.jetty.jsp.JettyJspServlet).  If there isn't a jsp servlet then the classname remains as null and an UnavailbleException as above is thrown (obviously without the preceding LOG.WARN as there is no conversion error).

Also, as has previously been the case, if DEBUG is enabled, there are more log messages available:

eg
 LOG.debug: Checking for precompiled servlet org.apache.jsp.x.blah_jsp for jsp /x/blah.jsp
if precompiled jsp is found:
     LOG.debug: JSP file /x/blah.jsp for foo mapped to Servlet org.apache.jsp.x.blah_jsp

if no precompiled jsp, but the jsp servlet is available:
    LOG.debug: JSP file /x/blah.jsp for foo mapped to JspServlet class org.eclipse.jetty.jsp.JettyJspServlet

I also changed the log messages for the case where the org.apache.jasper.compiler.JspUtil class (used for making jsp package and class names) is not present.:

Before:
LOG.warn: Unable to make package for jsp  /x/blah.jsp trying org.apache.jsp.x instead
LOG.warn: Unable to make identifier for jsp blah.jsp trying blah_jsp instead
Now:
jjava.lang.ClassNotFoundException: org.apache.jasper.compiler.JspUtil
LOG.debug: JspUtil.makeJavaPackage failed for /x/blah.jsp using org.apache.jsp.x instead
LOG.debug: JspUtil.makeJavaIdentifier failed for jsp blah.jsp using blah_jsp instead

I've also added some unit tests.

These changes are checked in to jetty-9.3.x, 9.4.x and master.
